language/diem-framework/modules/Vote.move

```diff
@@ -0,0 +1,377 @@
+        expiration_timestamp_secs: u64,
+    }
+
+    /// Ballots stores a list of ballots under a proposers address.
+    /// It is type parametrized by the Proposal. Some of these may
+    /// be expired and can be cleaned up using the `gc` function
+    /// It also contains the handles for all the events associated
+    /// with the ballots created by a proposer
+    struct Ballots<Proposal: store + drop> has key {
+        ballots: vector<Ballot<Proposal>>,
+        create_ballot_handle: Event::EventHandle<CreateBallotEvent<Proposal>>,
+        remove_ballot_handle: Event::EventHandle<RemoveBallotEvent>,
+        voted_handle: Event::EventHandle<VotedEvent>,
+        ballot_approved_handle: Event::EventHandle<BallotApprovedEvent>,
+    }
+
+    /// A counter which is stored under the proposers address and gets incremented
+    /// everytime they create a new ballot. This is used for creating unique
+    /// global identifiers for Ballots
+    struct BallotCounter has key {
+        counter: u64,
+    }
+
+    /// CreateBallotEvent is emitted when a ballot is
+    /// created by a proposer
+    struct CreateBallotEvent<Proposal: store + drop> has drop, store {
+        ballot_id: BallotID,
+        ballot: Ballot<Proposal>,
+    }
+
+    /// RemoveBallotEvent is emitted when a ballot has
+    /// been removed. This can either happen because:
+    /// * ballot was approved
+    /// * ballot was manually removed by the proposer
+    /// * ballot was expired and garbage collected
+    struct RemoveBallotEvent has drop, store {
+        ballot_id: BallotID,
+    }
+
+    /// VotedEvent is emitted when a valid vote has
+    /// been accepted by the ballot
+    struct VotedEvent has drop, store {
+        ballot_id: BallotID,
+        voter: address,
+        vote_weight: u64,
+    }
+
+    /// BallotApprovedEvent is emitted when a ballot has
+    /// been approved by the voters
+    struct BallotApprovedEvent has drop, store {
+        ballot_id: BallotID,
+    }
+
+    /// The maximum number of ballots allowed per proposal type
+    /// per address.
+    const MAX_BALLOTS_PER_PROPOSAL_TYPE_PER_ADDRESS: u64 = 256;
+
+    /// The provided timestamp(s) were invalid
+    const EINVALID_TIMESTAMP: u64 = 1;
+    /// The address already contains has the maximum of ballots allowed
+    /// MAX_BALLOTS_PER_PROPOSAL_TYPE_PER_ADDRESS
+    const ETOO_MANY_BALLOTS: u64 = 2;
+    /// Ballot with the provided id was not found
+    const EBALLOT_NOT_FOUND: u64 = 3;
+    /// Proposal details in the vote do not match the proposal details
+    /// in the ballot
+    const EBALLOT_PROPOSAL_MISMATCH: u64 = 4;
+    /// Voter not allowed to vote in the ballot
+    const EINVALID_VOTER: u64 = 5;
+    /// current timestamp > ballot expiration time
+    const EBALLOT_EXPIRED: u64 = 7;
+    /// Voter has already voted in this ballot
+    const EALREADY_VOTED: u64 = 8;
+
+    /// A constructor for BallotID
+    public fun new_ballot_id(
```

Comment:
Should this be public? Seems like it probably shouldn't...

If you need it for testing code you can make this function `public(friend)` and then add a friend annotation in this module just below the `use`s:

```rust
#[test_only]
friend DiemFramework::VoteTests;
```

Reply:
Don't need this for test. The `fun vote()` takes BallotID as a parameter and a client of this module would need to pass this struct when calling the `vote` function and hence needs to construct it.